0022-compartmentalize-grains-via-pop.md

```diff
@@ -0,0 +1,74 @@
+to be independently maintained.
+
+Working group leaders and other community members focused on specific platforms can manage the release cycle for
+individual idem-platform projects independently from salt.
+
+# Design
+[design]: #detailed-design
+
+Grains are currently created by calling functions in the salt loader.  We would simply need to call
+external POP functions to collect grains.  We need to maintain backwards compatibility with custom user grains.
+
+- All salt grains have already been ported to `grainsv2`.
+- `grainsv2` has many more features than the salt-grains implementation, but is compatible with salt grains.
+- None of the grain names have been changed, but the output of grains is more consistent/complete between platforms.
+- This will resolve many bugs that are open against grains and will migrate those bugs OUT of the salt
+platform -- making salt easier to develop
+- Grains in the new platform are independently fully tested.
+- Development will freeze on core.py and the salt/grains folder.
+- This comprises a complete rewrite of ALL grains.  Backwards compatibility is expected, but we need this to get
+in early so that any unintentional issues can be reet up a conversation with the relevant parties solved quickly.
+- Salt releases will aggressively pin `grainsv2`/idem-platform dependencies with each release.
+- Users will be to update their grains implementation to a custom idem-platform version.
+- The use of `grainsv2` will be gated with a config variable; It will be disabled by default.
+- New development for grains will be done in `grainsv2` projects and core.py will remain static.
+- When `grainsv2` gains maturity in salt on the other side of the gate (and python3.5 is dropped) it will completely replace core.py
+
+## Alternatives
+[alternatives]: #alternatives
+
+Keep things the way they are.  Does this mean we would suffer in stagnation?  Will we be stuck maintaining a monolith.
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+What will further adoption of POP mean inside of salt?  The purpose of this sep and these features is not to provide
+a definitive answer to this question -- rather to introduce plugin oriented programming into salt in a very safe way.
+In the future this will determine the design so that salt can be further compartmentalized and written in POP.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this? Please consider:
+
+- We will be introducing a number of new dependencies.
+- The new implementation of grains/pop is written with python3.6 and later with asyncio
+(The latest release of salt is also python3.6 and later).
+- Some grains for specific OSes (such as SUSE) implement some grains differently from other platforms.  We need
+to set up a conversation with the relevant parties to resolve such conflicts.
+- Documentation will be the heaviest lifting.  We need to document how to make salt-style internal grains and generally
```

Comment:
Specifically what documentation are we lacking now. What is the plan and time-lines around getting all of the needed documentation in place?

Reply:
For grains there isn't much more documentation changes that will need to happen, but for the future of the compartmentalization of salt this is an unaddressed problem.    Execution and state modules could be maintained in platform specific projects...  How do we merge that documentation into salt?